#Pipeline to detect salep ITS sequences from Oxford Nanopore Reads
#This is a three step workflow
#    1. Basecalling from fast5 or pod5 files to fastq files (This step can be skipped when fastq files are already calculated) [epi2me-labs/wf-basecalling]
#    2. Demultiplexing the sequences. A custom barcode file (FASTA and Arrangement File) has to be made and put into this workflow. [dorado]
#    3. Mapping and Counting of ITS sequences and assigning them to a species. [epi2me-labs/wf-metagenomics]

configfile: "../cfg/config.yaml"

#these are the standard demux and metagenomics input when basecalling, demux and metagenomics is done.

demux_input = "results/fastq/SAMPLE.pass.fq.gz"
metagenomics_input = "results/demux/barcoding_summary.txt" 

#if you want to skip demux set the input of metagenomics as the output of basecalling
if config["skip_demux"] == "YES":
    metagenomics_input = "results/fastq/SAMPLE.pass.fq.gz"
elif config["skip_demux"] == "NO":
    metagenomics_input = metagenomics_input
else:
    print("ERROR: skip_demux has to be set to either YES or NO")


#if already basecalled input is given set either the demux or metagenomics input as the fastq folder.
if config["nseq_input"] != "":
    if config["skip_demux"] == "YES":
        metagenomics_input = config["nseq_input"]
    elif config["skip_demux"] == "NO":
        demux_input = config["nseq_input"]
    else:
        print("ERROR: skip_demux has to be set to either YES or NO")

#rules
rule all:
    input:
        "results/metagenomics/wf-metagenomics-report.html"
    shell:
        "mv .nextflow.log* results/ && "
        "cp -r ../cfg results/"
 #folder clean up to have everything nicely packaged in results folder


rule basecall:
    input: config["pod5_input"]
    output:
        "results/fastq/SAMPLE.pass.fq.gz"
    params:
        basecaller_model=config["basecaller_model"],
        qscore_filter=config["qscore_filter"]
    conda: "conda_env.yaml"
    shell:
        "mkdir -p results && "
        "mkdir -p results/fastq && "
        "nextflow run epi2me-labs/wf-basecalling "
        "--input {input} "
        "--out_dir results/fastq "
        "--output_fmt fastq "
        "--qscore_filter {params.qscore_filter} "
        "--basecaller_cfg {params.basecaller_model} "
        "--cuda_device auto "
        "-profile singularity"

rule demux:
    input: demux_input
    output: "results/demux/barcoding_summary.txt"
    conda: "conda_env.yaml"
    shell:
        "mkdir -p results/demux && "
        "singularity run docker://nanoporetech/dorado dorado demux "
        "--barcode-arrangement ../cfg/barcodes_arrangement.toml " #custom barcode files
        "--barcode-sequences ../cfg/barcodes.fasta "
        "--sample-sheet ../cfg/sample_sheet.csv "
        "--kit-name \"SALEPBC\" " # don't change this name
        "--output-dir results/demux "
        "--emit-summary "
        "{input} && "
        "cd results/demux && " #create a folder for each sample and place the fastq in it
        "for f in ./*bam; "
        "do "
        "t=${{f%.*}} && "
        "T=${{t#*_}} && "
        "mkdir \"$T\" && "
        "mv \"$f\" \"$T\"; "
        "done && "
        "cd ../.."

#This is the command which is executed in the rule metagenomics
metagenomics_shell_command = "mkdir -p results/metagenomics && nextflow run epi2me-labs/wf-metagenomics --out_dir ./results/metagenomics --classifier minimap2 --reference ./references/Orchids_NCBI_PLANiTS_ITS_matched_to_ref_O_N_P.fasta --ref2taxid ./references/ref2taxid.O_N_P.tsv --taxonomy ./references/new_taxdump.tar.gz --include_read_assignments -profile singularity "

#if we want to skip demux and already basecalled files are given we need to add the input folder/file to the above command along with it's type (bam or fastq)
if config["skip_demux"] == "YES":
    if config["nseq_input"] != "":
        metagenomics_shell_command += "--" + config["nseq_format"] + " " + config["nseq_input"] + " "
    else:
        metagenomics_shell_command += "--fastq results/fastq/SAMPLE.pass.fq.gz " 
else:
    metagenomics_shell_command += "--bam ./results/demux "


rule metagenomics:
    input: metagenomics_input
    output: "results/metagenomics/wf-metagenomics-report.html"
    conda: "conda_env.yaml"
    shell:
        metagenomics_shell_command
